Further mitigate Undici ClientDestroyedError and improve HTTP resilience

parachemist opened the pull request:
## Multi-layer defense strategy for bulletproof HTTP

The application was experiencing multiple critical issues that caused crashes, errors, and operational instability:

### 1. **ClientDestroyedError Crashes**
```
[Error [ClientDestroyedError]: The client is destroyed] {
  code: 'UND_ERR_DESTROYED'
}
⨯ unhandledRejection: [Error [ClientDestroyedError]: The client is destroyed]
```
### 2. **Timeout Retry Loops**
```
Retry attempt 1/6 after 167ms [GET /users/...] - Error: This operation was aborted
Retry attempt 2/6 after 421ms [GET /users/...] - Error: This operation was aborted
[Hundreds of retries filling logs]
```
### 3. **Socket/Network Errors Failing Fast**
```
Non-retryable error [GET /pool/pool.status] - fetch failed
[cause]: SocketError: other side closed
  code: 'UND_ERR_SOCKET'
```
### 4. **Hard-coded Endpoints Causing 404s**
```
Non-retryable error [GET /pool/users] - Failed to fetch pool users:
Error collecting user stats: HttpError: Failed to fetch pool users:
  status: 404
```
### 5. **Duplicate Path Issues**
```
Fetched https://alpha.parasite.dev/aggregator/aggregator/pool/pool.status - Status: 404
```
### 6. **Invalid Data Causing Discovery Failures**
```
Could not find user stats endpoint. Tried: /aggregator/users/[], /users/[], /api/users/[]
```

These errors manifested in several ways:
- **Unhandled promise rejections** causing processes to crash
- **Truncated HTTP responses** landing `SyntaxError: Unexpected end of JSON input`
- **No automatic recovery** likely to blame from transient connection issues
- **Race conditions** at time of shutdown due to requests still in flight
- **Configuration incompatibility** across differentiating API setups
- **Invalid data** still being processed as valid addresses

## The solutions at hand

In this PR I will attempt to implement a comprehensive multi-layer defense strategy addressing the following vector of issues:

1. **Automatic retry logic** for the purpose of transparently handling transient errors with intelligent classification
2. **Request lifecycle tracking** in order to prevent premature connection destruction
3. **Graceful shutdown procedures** will ensure clean teardown with active request monitoring
4. **Circuit breaker pattern** cuts down on cascading failures and API hammering
5. **Endpoint discovery** finds correct API paths for any configuration at a whim
6. **Response format detection** for handling JSON arrays and newline-delimited text
7. **Address validation** now filters invalid data to prevent processing errors
8. **Socket error handling** correctly retries network-level failures
9. **Defensive parsing** takes care of malformed responses gracefully

## The changes that were made

 - `lib/http-client.ts` 
  - New retry logic, lifecycle tracking and graceful shutdown handling. 
 - `scripts/start-jobs.ts`
  - Shutdown and process safety enhanced.
 - `lib/pool-stats-collector.ts`
  - Error classification, endpoint discovery and validation.
 - `app/api/pool-stats/route.ts`
  - Aforementioned circuit breaker as well as caching and discovery.

eztvdpqx5 commented on app/api/pool-stats/route.ts:
```diff
@@ -15,34 +75,162 @@ export async function GET() {
     
-    // Combine the data from all three objects
-    const [statsData, hashrateData, diffData] = jsonLines;
+    for (let attempt = 0; attempt <= maxRetries; attempt++) {
```
do we want a retry here?  I thought the fetch from http-client already has retries but maybe this is retrying for some different state/situation?

parachemist commented:
Ready for re-review @paratoxicdev all should work as expected now.

On a side note, I'm not pleased with how the `pool-stats-collector.ts` has come out to 1100 lines. I am thinking that in my next PR I will clean it up a bit by extracting retry logic & types and making their own scripts in `/lib` for reuse in things like the `http-client` so we can just import and don't have to reuse logic. Will also spend some time consolidating the discovery functions. This will shave off about 300-400 lines.

paratoxicdev commented:
I'm still seeing this:

✓ Discovered pool stats endpoint: /aggregator/pool/pool.status
✓ Discovered pool users endpoint: /users
User lifecycle: 0 added, 2589 deactivated, 0 reactivated
Collected stats for 0 users at 2025-11-14T18:51:29.161Z
Pool stats collected and stored at 2025-11-14T18:51:29.294Z


your logic seems to be deactivting all users and stops collecting stats. Also the leaderboard doesn't show, also top difficulties. Basically nothing is loading.

paratoxicdev commented:
Hmm, what happens when you switch back to master then run the stuff, see if it works, then go back to your branch and run it?

parachemist commented:
Alright, after letting this branch run for a bit I also ran into a problem. I'm going to try messing with `isUserCollectorRunning`, I think that flag is hanging. My issue right now is that I get this spam:
```
⚠️  User stats collection already running, skipping this cycle
⚠️  User stats collection already running, skipping this cycle
⚠️  User stats collection already running, skipping this cycle
⚠️  User stats collection already running, skipping this cycle
⚠️  User stats collection already running, skipping this cycle
⚠️  User stats collection already running, skipping this cycle
```
As for the master branch, I get this output on `dev`:
```
 ✓ Starting...
 ✓ Ready in 1141ms
 ○ Compiling /api/pool-stats ...
 ✓ Compiled /api/pool-stats in 2.2s
Error fetching pool stats: SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at <unknown> (app/api/pool-stats/route.ts:25:64)
    at Array.map (<anonymous>)
    at GET (app/api/pool-stats/route.ts:25:47)
  23 |
  24 |     // Split the response into lines and parse each JSON object
> 25 |     const jsonLines = text.trim().split('\n').map(line => JSON.parse(line));
     |                                                                ^
  26 |
  27 |     // Combine the data from all three objects
  28 |     const [statsData, hashrateData, diffData] = jsonLines;
 GET /api/pool-stats 500 in 3266ms
 ⚠ Cross origin request detected from 192.168.4.84 to /_next/* resource. In a future major version of Next.js, you will need to explicitly configure "allowedDevOrigins" in next.config to allow this.
Read more: https://nextjs.org/docs/app/api-reference/config/next-config-js/allowedDevOrigins
Error fetching pool stats: SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at <unknown> (app/api/pool-stats/route.ts:25:64)
    at Array.map (<anonymous>)
    at GET (app/api/pool-stats/route.ts:25:47)
  23 |
  24 |     // Split the response into lines and parse each JSON object
> 25 |     const jsonLines = text.trim().split('\n').map(line => JSON.parse(line));
     |                                                                ^
  26 |
  27 |     // Combine the data from all three objects
  28 |     const [statsData, hashrateData, diffData] = jsonLines;
 GET /api/pool-stats 500 in 426ms
Error fetching pool stats: SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at <unknown> (app/api/pool-stats/route.ts:25:64)
    at Array.map (<anonymous>)
    at GET (app/api/pool-stats/route.ts:25:47)
  23 |
  24 |     // Split the response into lines and parse each JSON object
> 25 |     const jsonLines = text.trim().split('\n').map(line => JSON.parse(line));
     |                                                                ^
  26 |
  27 |     // Combine the data from all three objects
  28 |     const [statsData, hashrateData, diffData] = jsonLines;
 GET /api/pool-stats 500 in 387ms
Error fetching pool stats: SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at <unknown> (app/api/pool-stats/route.ts:25:64)
    at Array.map (<anonymous>)
    at GET (app/api/pool-stats/route.ts:25:47)
  23 |
  24 |     // Split the response into lines and parse each JSON object
> 25 |     const jsonLines = text.trim().split('\n').map(line => JSON.parse(line));
     |                                                                ^
  26 |
  27 |     // Combine the data from all three objects
  28 |     const [statsData, hashrateData, diffData] = jsonLines;
 GET /api/pool-stats 500 in 402ms
```
And as for the `collect-stats`:
```
> parastats@0.0.1 collect-stats /home/p/parastats
> tsx scripts/start-jobs.ts

Stats collectors started
📊 Pool stats collector started
👥 User auto-discovery enabled (runs with stats collection; set AUTO_DISCOVER_USERS=false to disable)
Attempting to connect to parasite.wtf:42069...
⚡ Stratum collector started
🚀 All jobs initialized and running
Non-retryable error [GET /pool/pool.status] - Failed to fetch pool stats:
Error collecting pool stats: HttpError: Failed to fetch pool stats:
    at <anonymous> (/home/p/parastats/lib/pool-stats-collector.ts:569:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async withRetry (/home/p/parastats/lib/pool-stats-collector.ts:118:14)
    at async collectPoolStats (/home/p/parastats/lib/pool-stats-collector.ts:553:18) {
  status: 404
}
Connected to Parasite stratum pool
📋 Connected and subscribed to Parasite stratum pool
🔧 Extranonce1: e495a768, Extranonce2 size: 8
🎯 New block template: Job 688bc71d000c89e1 (2025-11-14T21:53:44.000Z)
🔐 Authorization successful - receiving mining notifications
Non-retryable error [GET /users] - Failed to fetch users:
Error fetching pool users: HttpError: Failed to fetch users:
    at <anonymous> (/home/p/parastats/lib/pool-stats-collector.ts:369:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async withRetry (/home/p/parastats/lib/pool-stats-collector.ts:118:14)
    at async collectAllUserStats (/home/p/parastats/lib/pool-stats-collector.ts:356:19) {
  status: 404
}
⚠️  Pool users API returned empty - either API failed or pool is actually empty
Skipping auto-discovery and lifecycle management, collecting stats for known active users only
Collected stats for 0 known users at 2025-11-14T21:53:46.962Z
Non-retryable error [GET /pool/pool.status] - Failed to fetch pool stats:
Error collecting pool stats: HttpError: Failed to fetch pool stats:
    at <anonymous> (/home/p/parastats/lib/pool-stats-collector.ts:569:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async withRetry (/home/p/parastats/lib/pool-stats-collector.ts:118:14)
    at async collectPoolStats (/home/p/parastats/lib/pool-stats-collector.ts:553:18)
    at async Task._execution (/home/p/parastats/lib/pool-stats-collector.ts:628:5) {
  status: 404
}
Non-retryable error [GET /users] - Failed to fetch users:
Error fetching pool users: HttpError: Failed to fetch users:
    at <anonymous> (/home/p/parastats/lib/pool-stats-collector.ts:369:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async withRetry (/home/p/parastats/lib/pool-stats-collector.ts:118:14)
    at async collectAllUserStats (/home/p/parastats/lib/pool-stats-collector.ts:356:19)
    at async Task._execution (/home/p/parastats/lib/pool-stats-collector.ts:633:5) {
  status: 404
}
⚠️  Pool users API returned empty - either API failed or pool is actually empty
Skipping auto-discovery and lifecycle management, collecting stats for known active users only
Collected stats for 0 known users at 2025-11-14T21:54:00.748Z
```
The JSON input error is quite prevalent, this and the `Non-retryable GET`s appear to be fixed on my branch. The master branch won't even start to collect stats on my instance, I just cloned it a few minutes ago with `git clone https://github.com/parasitepool/parastats.git`.

Either way like I said, pretty sure that flag is hanging. Tinkering now.

paratoxicdev reviewed (changes requested):
I'm still getting this: 

❌ Pool users list is EMPTY! API may have returned invalid data. Skipping cycle to prevent deactivating all users.
Raw API response (first 500 chars): []

On the master branch everything works fine as it should. I think you need to radically simplify this PR to figure out what is going on. It feels like the LLM just added a bunch of unnecessary complexity and now it's difficult to find the bug

parachemist commented:
> I'm still getting this:
> 
> ❌ Pool users list is EMPTY! API may have returned invalid data. Skipping cycle to prevent deactivating all users. Raw API response (first 500 chars): []
> 
> On the master branch everything works fine as it should. I think you need to radically simplify this PR to figure out what is going on. It feels like the LLM just added a bunch of unnecessary complexity and now it's difficult to find the bug

Alright I will go ahead and start fresh from master and try use what I can from here. Quite odd as to why it works on my machine (ThinkPad P50) and not your Para laptop.
